src/ConfigurationClientManager.ts

```diff
@@ -0,0 +1,332 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { AppConfigurationClient, AppConfigurationClientOptions } from "@azure/app-configuration";
+import { ConfigurationClientWrapper } from "./ConfigurationClientWrapper";
+import { TokenCredential } from "@azure/identity";
+import { AzureAppConfigurationOptions, MaxRetries, MaxRetryDelayInMs } from "./AzureAppConfigurationOptions";
+import { isFailoverableEnv } from "./requestTracing/utils";
+import * as RequestTracing from "./requestTracing/constants";
+
+const TCP_ORIGIN = "_origin._tcp";
+const ALT = "_alt";
+const TCP = "_tcp";
+const EndpointSection = "Endpoint";
+const IdSection = "Id";
+const SecretSection = "Secret";
+const AzConfigDomainLabel = ".azconfig.";
+const AppConfigDomainLabel = ".appconfig.";
+const FallbackClientRefreshExpireInterval = 60 * 60 * 1000; // 1 hour in milliseconds
+const MinimalClientRefreshInterval = 30 * 1000; // 30 seconds in milliseconds
+const SrvQueryTimeout = 5000; // 5 seconds
+
+interface IConfigurationClientManager {
+    getClients(): Promise<ConfigurationClientWrapper[]>;
+    refreshClients(): Promise<void>;
+}
+
+export class ConfigurationClientManager implements IConfigurationClientManager {
+    isFailoverable: boolean;
+    endpoint: string;
+    #secret : string;
+    #id : string;
+    #credential: TokenCredential;
+    #clientOptions: AppConfigurationClientOptions | undefined;
+    #validDomain: string;
+    #staticClients: ConfigurationClientWrapper[];
+    #dynamicClients: ConfigurationClientWrapper[];
+    #lastFallbackClientRefreshTime: number = 0;
+    #lastFallbackClientRefreshAttempt: number = 0;
+
+    constructor (
+        connectionStringOrEndpoint?: string | URL,
+        credentialOrOptions?: TokenCredential | AzureAppConfigurationOptions,
+        appConfigOptions?: AzureAppConfigurationOptions
+    ) {
```

Comment:
Could we use constructor overload here? I know this capability of TS that can help flexibly pass arguments, but it looks too hacky to input different set of parameters.

Reply:
There is no overload in javascript. The best we can do is something like [this](https://github.com/Azure/AppConfiguration-JavaScriptProvider/blob/main/src/load.ts#L28) But i am also no sure whether it is possible for constructor @Eskibear